### PR TITLE
Fixed #22341 -- Turned django.db.models.fields.related into a package

### PR DESCRIPTION
**Pull request by loic**
**POC** turned into a PR for **review purpose**.

**Not ready for commit.**

### CODE REVIEW

File: django/db/models/fields/__init__.py [hunk 1]
```diff
@@ -365,14 +365,21 @@ def deconstruct(self):
...
+            "django.db.models.fields.related.many_to_one",
+            "django.db.models.fields.related.many_to_many",
+            "django.db.models.fields.related.one_to_one",
+            "django.db.models.fields.related",
+            "django.db.models.fields.files",
+            "django.db.models.fields.proxy",
+            "django.db.models.fields",
+        ]
+        for flatten_path in flatten_paths:
+            if path.startswith(flatten_path):
```
Comment (timgraham) [inline django/db/models/fields/__init__.py]: maybe remove the `django.db.models.fields` repetition above and use `path.startswith("django.db.models.fields" + flatten_path)` here?

File: django/db/models/fields/related/base.py [hunk 2]
```diff
@@ -0,0 +1,1115 @@
...
+from django.core import checks
+from django.db import connections, router, transaction
+from django.db.models import signals
+from django.db.models.deletion import CASCADE
+from django.db.models.fields import BLANK_CHOICE_DASH, Field, FieldDoesNotExist
+from django.db.models.lookups import IsNull
+from django.db.models.query_utils import PathInfo
+from django.db.models.expressions import Col
+from django.utils.encoding import force_text, smart_text
+from django.utils import six
```
Comment (timgraham) [inline django/db/models/fields/related/base.py]: order this above `.encoding`

File: django/db/models/fields/related/many_to_one.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
+from django import forms
+from django.core import checks, exceptions
+from django.db import router
+from django.db.models.fields import AutoField, FieldDoesNotExist, IntegerField, PositiveIntegerField, PositiveSmallIntegerField
+from django.db.models.deletion import CASCADE, SET_NULL, SET_DEFAULT
```
Comment (timgraham) [inline django/db/models/fields/related/many_to_one.py]: above `.fields`

File: django/db/models/fields/related/many_to_one.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
+from django import forms
+from django.core import checks, exceptions
+from django.db import router
+from django.db.models.fields import AutoField, FieldDoesNotExist, IntegerField, PositiveIntegerField, PositiveSmallIntegerField
```
Comment (timgraham) [inline django/db/models/fields/related/many_to_one.py]: multiline?

**Comment by timgraham** [comment]
Looks good to me. I guess you'll have to duke it out with @PirosB3 and @freakboy3742 as to who wants to merge first.

**Comment by PirosB3** [comment]
I'll have a look as soon as I come back

**Comment by loic** [comment]
@timgraham @PirosB3, I think we are going to postpone this refactor some more. After discussing this with @akaariai again we both agree that more work is needed to decouple FO and FK properly, we should also fix the inconsistent and confusing naming of classes. Finally we want to experiment with making reverse relation actual fields.

It's a grand plan, but if we can't pull it off we can always do an unambitious split up later. Also I suspect Daniel's Meta refactor can help quite a bit.

**Comment by PirosB3** [comment]
Hi all,
@loic I like the new structuring a lot. related.py is too big currently.
@loic, once the _meta API is in I would love to help you in this. By doing the _meta API I have seen many parts of the code that should be redesigned once the new API is in.

**Comment by timgraham** [comment]
Okay, I assume this won't happen for 1.8 alpha, so I'll remove that tag from the ticket.

**Pull request closed without merging**